crates/solver-service/src/server.rs

```diff
@@ -507,3 +543,125 @@ async fn forward_to_discovery_service(
 		},
 	}
 }
+
+/// Calculates the profit margin percentage for an order.
+///
+/// The profit margin is calculated as:
+/// Profit = Total Input Amount (USD) - Total Output Amount (USD) - Execution Costs (USD)
+/// Profit Margin = (Profit / Total Input Amount (USD)) * 100
+///
+/// This represents the percentage profit the solver makes on the input amount.
+/// All amounts are converted to USD using the pricing service for accurate comparison.
+async fn calculate_order_profitability(
+	order: &Order,
+	cost_estimate: &CostEstimate,
+	solver: &SolverEngine,
+) -> Result<Decimal, Box<dyn std::error::Error>> {
+	// Parse the order data as Eip7683OrderData
+	let order_data: solver_types::Eip7683OrderData = serde_json::from_value(order.data.clone())
+		.map_err(|e| format!("Failed to parse order data: {}", e))?;
+
+	let token_manager = solver.token_manager();
+	let pricing_service = solver.pricing();
+
+	// Calculate total input amount in USD (sum of all inputs converted to USD)
+	let mut total_input_amount_usd = Decimal::ZERO;
+	for input in &order_data.inputs {
```

Comment:
This type of parsing is again very specific to 7683 which has the ("inputs") field. Other order types might not have this, or it may be called differently..